app/hooks/useNetworkState.tsx

```diff
@@ -0,0 +1,28 @@
+import { useEffect, useRef, useState } from "react";
+
+export function useNetworkState() {
+    const [isOnline, setIsOnline] = useState(window.navigator.onLine);
+    const callbackRef = useRef<() => void>();
+
+    const on = () => {
+        setIsOnline(true);
+        callbackRef.current && callbackRef.current();
+    };
+    const off = () => setIsOnline(false);
```

Comment:
These callbacks should've been defined inside of `useEffect` to avoid stale closure side effects